protos/src/auth_service/convert.rs

```diff
@@ -0,0 +1,397 @@
+};
+
+impl From<identifiers::AsClientId> for AsClientId {
+    fn from(value: identifiers::AsClientId) -> Self {
+        let (user_name, client_id) = value.into_parts();
+        Self {
+            user_name: Some(user_name.into()),
+            client_id: Some(client_id.into()),
+        }
+    }
+}
+
+impl TryFrom<AsClientId> for identifiers::AsClientId {
+    type Error = AsClientIdError;
+
+    fn try_from(proto: AsClientId) -> Result<Self, Self::Error> {
+        Ok(Self::new(
+            proto
+                .user_name
+                .ok_or_missing_field("user_name")?
+                .try_into()?,
+            proto.client_id.ok_or_missing_field("client_id")?.into(),
+        ))
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum AsClientIdError {
+    #[error(transparent)]
+    MissingField(#[from] MissingFieldError<&'static str>),
+    #[error(transparent)]
+    QualifiedUserName(#[from] QualifiedUserNameError),
+}
+
+impl From<AsClientIdError> for Status {
+    fn from(e: AsClientIdError) -> Self {
+        Status::invalid_argument(format!("invalid AS client id: {e}"))
+    }
+}
+
+impl From<credentials::ClientCredentialCsr> for ClientCredentialCsr {
+    fn from(value: credentials::ClientCredentialCsr) -> Self {
+        Self {
+            msl_version: value.version as u32,
+            client_id: Some(value.client_id.into()),
+            signature_scheme: value.signature_scheme as i32,
+            verifying_key: Some(value.verifying_key.into()),
+        }
+    }
+}
+
+impl TryFrom<ClientCredentialCsr> for credentials::ClientCredentialCsr {
+    type Error = ClientCredentialCsrError;
+
+    fn try_from(proto: ClientCredentialCsr) -> Result<Self, Self::Error> {
+        let version = match proto.msl_version {
+            0 => messages::MlsInfraVersion::Alpha,
+            version => return Err(ClientCredentialCsrError::InvalidMlsVersion(version)),
+        };
+        let signature_scheme = SignatureScheme::try_from(proto.signature_scheme)
+            .map_err(|_| UnsupportedSignatureScheme)?
+            .try_into()?;
+
+        Ok(Self {
+            version,
+            client_id: proto
+                .client_id
+                .ok_or_missing_field("client_id")?
+                .try_into()?,
+            signature_scheme,
+            verifying_key: proto
+                .verifying_key
+                .ok_or_missing_field("verifying_key")?
+                .into(),
+        })
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ClientCredentialCsrError {
+    #[error("invalid MLS version: {0}")]
+    InvalidMlsVersion(u32),
+    #[error(transparent)]
+    Field(#[from] MissingFieldError<&'static str>),
+    #[error(transparent)]
+    Signature(#[from] UnsupportedSignatureScheme),
+    #[error(transparent)]
+    ClientId(#[from] AsClientIdError),
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("unsupported MLS version: {0}")]
+pub struct UnsupportedMlsVersion(u32);
```

Comment:
Is this really about the MLS version? Or about the version of our protocol?

Reply:
Good catch. It is actually about the mls version being "unsupported" or even better "unexpected". I renamed the error type.